extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java

```diff
@@ -143,147 +123,161 @@ public void resume(TaskLocation location, String id)
-
+      final FullResponseHolder response = submitRequest(id, HttpMethod.GET, "offsets/end", null, true);
       return jsonMapper.readValue(response.getContent(), new TypeReference<Map<Integer, Long>>() {});
     }
-    catch (Exception e) {
+    catch (IOException e) {
       throw Throwables.propagate(e);
     }
   }
 
-  public void setEndOffsets(TaskLocation location, String id, Map<Integer, Long> endOffsets)
+  public void setEndOffsets(String id, Map<Integer, Long> endOffsets)
   {
-    setEndOffsets(location, id, endOffsets, false);
+    setEndOffsets(id, endOffsets, false);
   }
 
-  public void setEndOffsets(TaskLocation location, String id, Map<Integer, Long> endOffsets, boolean resume)
+  public void setEndOffsets(String id, Map<Integer, Long> endOffsets, boolean resume)
   {
     try {
-      final URI serviceUri = new URI(
-          "http",
-          null,
-          location.getHost(),
-          location.getPort(),
-          String.format("%s/%s/%s", BASE_PATH, id, "offsets/end"),
-          resume ? "resume=true" : null,
-          null
-      );
       submitRequest(
-          new Request(HttpMethod.POST, serviceUri.toURL()).setContent(
-              MediaType.APPLICATION_JSON,
-              jsonMapper.writeValueAsBytes(endOffsets)
-          ), true
+          id,
+          HttpMethod.POST,
+          "offsets/end",
+          resume ? "resume=true" : null,
+          jsonMapper.writeValueAsBytes(endOffsets),
+          true
       );
     }
-    catch (Exception e) {
+    catch (IOException e) {
       throw Throwables.propagate(e);
     }
   }
 
-  private StatusResponseHolder submitRequest(Request request, boolean retry)
+  @VisibleForTesting
+  RetryPolicyFactory createRetryPolicyFactory()
+  {
+    // Retries for about a minute before giving up; this should be long enough to handle any temporary unresponsiveness
+    // such as network issues, if a task is still in the process of starting up, or if the task is in the middle of
+    // persisting to disk and doesn't respond immediately.
+
+    return new RetryPolicyFactory(
+        new RetryPolicyConfig()
+            .setMinWait(Period.seconds(2))
+            .setMaxWait(Period.seconds(8))
+            .setMaxRetryCount(8)
+    );
+  }
+
+  @VisibleForTesting
+  void checkConnection(String host, int port) throws IOException
+  {
+    new Socket(host, port).close();
+  }
+
+  private FullResponseHolder submitRequest(String id, HttpMethod method, String pathSuffix, String query, boolean retry)
+  {
+    return submitRequest(id, method, pathSuffix, query, new byte[0], retry);
+  }
+
+  private FullResponseHolder submitRequest(
+      String id,
+      HttpMethod method,
+      String pathSuffix,
+      String query,
+      byte[] content,
+      boolean retry
+  )
   {
     final RetryPolicy retryPolicy = retryPolicyFactory.makeRetryPolicy();
     while (true) {
```

Comment:
It could, but the way the logic is written now it'll propagate the last exception when it runs out of retries which is desirable and would take more logic to implement if the loop could end in a way other than throwing or returning.